Mathlib/Lean/Meta/DiscrTree.lean

```diff
@@ -53,19 +53,25 @@ partial def getSubexpressionMatches (d : DiscrTree α s) (e : Expr) : MetaM (Arr
 
 variable {m : Type → Type} [Monad m]
 
-/-- Apply a monadic function to the array of values at each node in a `DiscrTree`. -/
-partial def Trie.mapArraysM (t : DiscrTree.Trie α s) (f : Array α → m (Array β)) :
-    m (DiscrTree.Trie β s) := do
-  match t with
-  | .node vs children =>
-    return .node (← f vs) (← children.mapM fun (k, t') => do pure (k, ← t'.mapArraysM f))
 
-/-- Apply a monadic function to the array of values at each node in a `DiscrTree`. -/
-def mapArraysM (d : DiscrTree α s) (f : Array α → m (Array β)) : m (DiscrTree β s) := do
-  pure { root := ← d.root.mapM (fun t => t.mapArraysM f) }
+/-- A representation of lists-of-quadruples with a smaller memory footprint. -/
+private inductive Ctxt (α β γ δ : Type)
+  | empty : Ctxt α β γ δ
+  | ctxt : α → β → γ → δ → Ctxt α β γ δ → Ctxt α β γ δ
+
+/-- Apply a function to the array of values at each node in a `DiscrTree`. -/
+partial def Trie.mapArrays (t : Trie α s) (f : Array α → Array β) : Trie β s :=
+  let .node vs0 cs0 := t
+  go Ctxt.empty #[] (f vs0) cs0.toList
+where
+  /-- This implementation as a single tail-recursive function is chosen to not blow the
+      interpreter stack when the `Trie` is very deep -/
+  go  | .empty, cs, vs, []                   => .node vs cs
+      | .ctxt cs vs todo k ps, cs', vs', []  => go ps (cs.push (k, .node vs' cs')) vs todo
+      | ps, cs, vs, (k, .node vs' cs')::todo => go (.ctxt cs vs todo k ps) #[] (f vs') cs'.toList
```

Comment:
the `toList` here is unnecessary, you can loop over the array with an index `i`.

Reply:
I had that before, but this led to prettier code, and is also necessary to allow it to free the memory as it goes.

Reply:
To free the memory, you can do this (similar to `Array.mapM`):
```lean
      let (k, .node vs' cs') := todo[i]
      -- clear elements from the array after visiting them
      let todo := todo.set ⟨i, h⟩ (unsafe unsafeCast (default : NonScalar))
      go (.ctxt cs vs todo (i+1) k ps) #[] (f vs') cs' 0
```
although whether this is actually a win depends on whether we expect the input `t` to be unshared or not. As for prettier code, this is generally not the main priority, especially for an optimization PR. The prettiest code is the one liner that was there originally. If we are going to optimize it, we may as well do it properly.

Reply:
I’d say it’s a hotfix PR to work around absurd trees, the real fix should be something like https://github.com/leanprover/lean4/pull/2577 :-)

`unsafe unsafCast` looks like a far too big hammer. I’ll use an array, but will simply reverse it and then `pop` elements off it. This frees memory, avoids lists, is in-place if the trie is exclusive and almost looks good. WDYT?

Reply:
How about this?

Reply:
(I’ll merge tomorrow; there are many ways to skin this cat, no need to dwell too long on the various variants.)